### ISSUE TITLE
`peerDependencies` interfering with update of `devDependencies` of same name

### ISSUE THREAD

**Issue by brettz9** [raineorshine/npm-check-updates#598]
- [x] I have read the list of [known issues](https://github.com/tjunnone/npm-check-updates#known-issues) before filing this issue.
- [x] I have searched for [similiar issues](https://github.com/tjunnone/npm-check-updates/issues) before filing this issue.

* node version: 10.16.3
* npm version: 6.12.0
* npm-check-updates version: 3.1.25

---------------------------

On attempting to update the package `eslint-plugin-chai-friendly` (see https://github.com/ihordiachenko/eslint-plugin-chai-friendly/blob/master/package.json ), the presence of its `peerDependencies` seems to prevent the same (`eslint`) devDep. being found for upgrade in `ncu -u`

**Comment by brettz9** [comment]
I think the reason why `dependencies` (or actually `devDependencies`, as I wouldn't expect `dependencies` to co-exist with `peerDependencies` since the purpose of the latter is to indicate a package should load the peer deps. itself) takes preference is because one may wish to support older versions (i.e., in `peerDependencies`), but when performing one's tests, use the most up-to-date version in `devDependencies`.

In this case, we let people use older versions of eslint if they wish (esp. since it can be troublesome to upgrade), but ensure we are testing with the latest version ourselves.

**Comment by mikestopcontinues** [comment]
I'm having this issue as well. I need to test my package with devDependencies, but when I deploy, I need end-users to use peerDependencies. I think they should definitely be handled separately.

**Comment by bugzpodder** [comment]
same here, devDependencies is needed for local testing.  i have peerDepenedencies at package: ^2.0.0 and devDepenedencies at package: 2.3.1 and ncu can't handle this.

**Comment by brettz9** [comment]
Note that as a workaround, one can do something like this in `.ncurc.js`:

```js
module.exports = {
  // Whitelist all for checking besides `peer`:
  "dep": "prod,dev,optional,bundle"
};
```

**Comment by mesqueeb** [comment]
I would say, since `peerDependencies` are not installed in your node_modules, you still need to add it to the `devDependencies` because you'll probably need it to run your tests.

So I think in 90% of the cases, people would have the same packages in `peer` and `dev` dependencies.

I'd like to keep them the same version at all times, when I upgrade one, also upgrade the other. Just to prevent anything breaking, because not everyone keeps to semver...

**Comment by brettz9** [comment]
@mesqueeb : While one should install in `devDependencies`, allowing lower versions (as well as higher version) in `peerDependencies` allows your consumers to keep using older versions.

For example, we do not want to give warnings to everyone using our eslint plugin to upgrade to the latest version of our peerDep, eslint immediately (by forcing `peerDependencies` to the latest eslint version), but we do want to use the latest eslint ourselves (using Travis to conditionally downgrade when we also want to check older peerDep (eslint) versions).

I'd be ok with peerDeps. being auto-updated to *add* the latest version though (e.g., expanding a current range).

**Comment by dominikg** [comment]
> Note that as a workaround, one can do something like this in `.ncurc.js`:
> 
> ```js
> module.exports = {
>   // Whitelist all for checking besides `peer`:
>   "dep": "prod,dev,optional,bundle"
> };
> ```

even with this, the version in peerDependencies  is updated in package.json if the updated dependency is also a devDependency.

given `foo@latest === foo@1.0.1`
and this package.json
```js
{
  "peerDependencies": {
    "foo": "^1.0.0"
  },
 "devDependencies": {
   "foo": "^1.0.0"
  }
}
```
running ncu with `--dep dev,prod,optional,bundle` i'd expect
```diff
{
  "peerDependencies": {
    "foo": "^1.0.0"
  },
 "devDependencies": {
-  "foo": "^1.0.0"
+  "foo": "^1.0.1" 
  }
}
```
but get 
```diff
{
  "peerDependencies": {-
-  "foo": "^1.0.0"
+  "foo": "^1.0.1" 
  },
 "devDependencies": {
-  "foo": "^1.0.0"
+  "foo": "^1.0.1" 
  }
}
```

is there any way to prevent this, esp. if the peer range already covers the update?

**Comment by raineorshine** [comment]
@dominikg It is definitely a bug and should be fixed. I can't think of a workaround. The method that re-writes the package.json is being too heavy-handed and should limit its search-and-replace to the targeted sections.

**Comment by raineorshine** [comment]
In `v12.0.0`, peer dependencies are excluded by default (i.e. `--dep prod,dev,optional,bundle`). This is the workaround that @brettz9 [suggested above](https://github.com/raineorshine/npm-check-updates/issues/598#issuecomment-611279430).

This doesn't fix the bug, but I thought I would mention it since it is related.

**Comment by raineorshine** [comment]
There isn't anybody actively working on this, but I don't think it's that hard a change. Just need to scope the range of the replace in https://github.com/raineorshine/npm-check-updates/blob/main/src/lib/upgradePackageData.ts based on `options.dep`.

I am open to PR's, or I can take a look when I have some free time.

**Comment by raineorshine** [comment]
I had a chance to tackle it this morning. Now only the chosen dependency sections will be upgraded (by default, everything except `peerDependencies`).

Published in `v16.5.1`.

**Comment by Badisi** [comment]
I can confirm `v16.5.1` is fixing the issue described by @dominikg.

Thanks a lot @raineorshine !
